darwin runtime support

thehajime opened the pull request:
This PR tries to add new platform support with darwin, since we have OCI runtime (https://github.com/ukontainer/runu) running on macos (w/o hypervisor.framework).

The commits include several specific issues (thus draft PR); several are already addressed but I would like to ask opinions from more people if there are better ways to handle.

- reap process handling is tricky since there are no PR_SET_CHILD_SUBREAPER alike on darwin
- I'm completely unfamiliar with native snapshotter; thus the current patch (16845ac0b64362186673b6af5873f549904fea01) is a bandaid to pass `ctr run` without failures.
  - `make integration` fails right now, thus ignored for the moment (will revert it)
- go-runc package needs to be updated (thus skip validate/vendor for the moment; will revert as well)
  - fchown with anonymous pipes on darwin returns EINVAL
- skip critest since grpc.v1.cri for darwin is not ported yet; disabled the test as well

I have another patchset for dockerd: will work on when everything required is upstreamed to containerd.

AkihiroSuda commented on .github/workflows/ci.yml:
```diff
@@ -455,3 +455,99 @@ jobs:
+      - name: Install runtime
+        shell: bash
+        run: |
+          GO111MODULE=on go get -v github.com/ukontainer/runu
```
Please pin a specific version

thehajime replied:
addressed this temporary.  I will update with a proper version tag later.

thehajime replied:
done.

AkihiroSuda commented on cmd/containerd-shim-darwin-v1/main.go:
```diff
@@ -0,0 +1,31 @@
+)
+
+func main() {
+	shim.Run("io.containerd.darwin.v1", New)
```
I'd prefer this to be called `io.containerd.runu.v1` and maintained in runu's repo

AkihiroSuda replied:
Or just extend the existing `io.containerd.runc.v2` to support Darwin

thehajime replied:
agree; will move those parts (cmd/containerd-shim-darwin-v1) to runu repo.

thehajime replied:
did remove cmd/containerd-shim-darwin-v1.  I will update the runu part later.

ktock commented on mount/mount_darwin.go:
```diff
@@ -0,0 +1,45 @@
+	"os"
+)
+
+// Use symlink instead of union mount on darwin
+func (m *Mount) Mount(target string) error {
```
Do we have snapshotters natively work on darwin? Should we have a new one?

thehajime replied:
at least, make integration reported some of errors related to snapshotters (which I was struggled to interpret them).

https://github.com/containerd/containerd/runs/1064611094?check_suite_focus=true#step:10:13

It would be helpful if somebody let me know what I have to do for a proper snapshotter.

thehajime replied:
Because bind mount is faked with a symbolic link, some features are not available.  But the current integration test for the native snapshot looks okay to me (I disabled TestSnapshotterClient/ViewReadonly for darwin); how do you think ?

```
--- PASS: TestSnapshotterClient (0.01s)
    --- PASS: TestSnapshotterClient/Chown (1.07s)
    --- PASS: TestSnapshotterClient/Basic (1.11s)
    --- PASS: TestSnapshotterClient/RemoveDirectoryInLowerLayer (1.57s)
    --- PASS: TestSnapshotterClient/Remove (1.22s)
    --- PASS: TestSnapshotterClient/Walk (1.76s)
    --- PASS: TestSnapshotterClient/PreareViewFailingtest (0.89s)
    --- PASS: TestSnapshotterClient/TransitivityTest (0.56s)
    --- PASS: TestSnapshotterClient/StatComitted (0.23s)
    --- SKIP: TestSnapshotterClient/ViewReadonly (0.00s)
    --- PASS: TestSnapshotterClient/StatActive (0.16s)
    --- PASS: TestSnapshotterClient/RootPermission (0.16s)
    --- PASS: TestSnapshotterClient/CloseTwice (0.42s)
    --- PASS: TestSnapshotterClient/Update (3.24s)
    --- PASS: TestSnapshotterClient/StatInWalk (1.11s)
    --- SKIP: TestSnapshotterClient/Rename (0.00s)
    --- PASS: TestSnapshotterClient/MoveFileFromLowerLayer (1.00s)
    --- PASS: TestSnapshotterClient/DeletedFilesInChildSnapshot (1.76s)
    --- PASS: TestSnapshotterClient/RemoveIntermediateSnapshot (1.05s)
    --- PASS: TestSnapshotterClient/DirectoryPermissionOnCommit (1.37s)
    --- PASS: TestSnapshotterClient/LayerFileupdate (8.69s)
    --- PASS: TestSnapshotterClient/128LayersMount (21.91s)
```

dmcgowan replied:
This should check to make sure the mount is a bind mount before doing the symlink hack.

We should find a better snapshot solution here, what type of block device/file support does runu have here? That and the unpacking are what need to be figured out here to have something closer to a real solution.

thehajime replied:
>This should check to make sure the mount is a bind mount before doing the symlink hack.

Okay, I will add the check before the hack.

>We should find a better snapshot solution here, what type of block device/file support does runu have here? That and the unpacking are what need to be figured out here to have something closer to a real solution.

Currently, runu supports two modes for mounting filesystems:
- a disk image file, formatted with Linux-capable filesystems (ext4, btrfs, etc)
- 9pfs hosted by server (experimental, running inside runu runtime, still a bit buggy)

So, because applications (with LKL) running over runu mount filesystem by themselves, runtimes (containerd, runu) are required to expose those files or 9pfs server ports to offer file services.

I'm not well understanding what part of _unpacking_ is not a real solution right now.

dmcgowan replied:
The unpacking today is done on the host. If the host is unable to mount the snapshot filesystem, then the unpacking will need to be done inside the runtime environment. Using the native snapshotter and simulating bind mounts with symlinks is not a real solution that we would recommend to use, but gets us a test environment.

So I take it from your response that this is using the experimental 9pfs inside the runu runtime.

thehajime replied:
Thanks, I got the point of what the native snapshotter is.

A decent snapshotter for darwin would be nice rather than the one of testing purpose, but I would expect this to further contribution because it's not likely to be a trivial component.

Or, is not having a real snapshotter a blocker for this pull request ?

> So I take it from your response that this is using the experimental 9pfs inside the runu runtime.

Yes, plus a disk image file unpacked by the snapshotter also can be mounted.

samuelkarp replied:
Does the `native` snapshotter work here?  It'll be slower, but it's very simple.

thehajime replied:
> Does the `native` snapshotter work here? It'll be slower, but it's very simple.

Yes.  This PR uses native snapshotter indeed.  On darwin there is no bind mount nor nullfs (yes it actually has but depends on fuse/bindfs etc), thus emulates with symlinks.

thehajime commented:
Currently, I'm waiting for a merge https://github.com/containerd/go-runc/pull/64; once it's upstreamed, I'll update this PR too.

AkihiroSuda commented on .github/workflows/ci.yml:
```diff
@@ -123,7 +123,7 @@ jobs:
 
       - name: Vendor
-        run: ../project/script/validate/vendor
+        run: ../project/script/validate/vendor || true
```
Please revert this

thehajime replied:
Yes, I'll revert it after the go-runc patch is merged.

thehajime replied:
done.

AkihiroSuda commented on .github/workflows/ci.yml:
```diff
@@ -455,3 +455,111 @@ jobs:
+          export PATH=$PATH:$(go env GOPATH)/bin
+          # default GOPATH under sudo is $HOME/go, not $HOME/.go on macos
+          sudo -E make integration
+      - name: CRI test
```
Can we remove this?

thehajime replied:
Agree.  The current test do almost nothing.

thehajime replied:
done.

AkihiroSuda commented on container_test.go:
```diff
@@ -174,6 +178,10 @@ func TestContainerStart(t *testing.T) {
 	t.Parallel()
 
+	if runtime.GOOS == "darwin" {
+		t.Skip("test with echo isn't supported on darwin")
```
Why?

thehajime replied:
Linux binary execution from public images (elf/amd64) on runu runtime is experimental and still has several limitations (only vfork, etc).  Thus the tests only use static-binary version of executable, and this image doesn't have various binaries (e.g., coreutils, etc).

So we did't test such stuffs in this PR.  I think we can revert it once it supports well.

thehajime replied:
I found creating `client_darwin_test.go`  would reduce the patchsize while addressing darwin-specific behaviors of the tests.  Will try it.

thehajime replied:
I have investigated what darwin containerd can test with `container_test.go`, and found that some of tests (11 out of 28 tests) should be skipped.


```
--- SKIP: TestWaitStoppedProcess (0.00s)
--- SKIP: TestContainerCloseIO (0.00s)
--- SKIP: TestWaitStoppedTask (0.00s)
--- PASS: TestNewClient (0.00s)
--- PASS: TestContainerUpdate (0.35s)
--- SKIP: TestKillContainerDeletedByRunc (0.00s)
--- PASS: TestShortRunningTaskPid (1.67s)
--- PASS: TestContainerExecNoBinaryExists (2.02s)
--- SKIP: TestContainerHook (0.00s)
--- PASS: TestContainerKill (2.79s)
--- PASS: TestContainerNoBinaryExists (3.51s)
--- PASS: TestDeleteRunningContainer (2.02s)
--- SKIP: TestDeleteContainerExecCreated (0.00s)
--- PASS: TestContainerExtensions (0.55s)
--- PASS: TestContainerLabels (0.55s)
--- SKIP: TestProcessForceDelete (0.00s)
--- SKIP: TestContainerHostname (0.00s)
--- PASS: TestContainerInfo (0.30s)
--- PASS: TestContainerExitedAtSet (1.40s)
--- SKIP: TestContainerLargeExecArgs (0.00s)
--- SKIP: TestContainerExec (0.00s)
--- PASS: TestNewContainer (0.55s)
--- SKIP: TestContainerStart (0.00s)
--- PASS: TestContainerPids (1.70s)
--- PASS: TestClientReconnect (0.00s)
--- PASS: TestContainerOutput (1.91s)
--- PASS: TestTaskForceDelete (1.57s)
--- PASS: TestShimSockLength (7.50s)
```


Main obstacle is the containerd-shim can't report the exit status due to different handling of reaper on darwin (i.e., we cannot use `wait4(2)` as processes aren't children without PR_SET_CHILD_SUBREAPER support).

Thus, I'm going to include limited tests in this PR but will update later once the shim (and runtimes) can report the status.

AkihiroSuda commented on container_test.go:
```diff
@@ -107,6 +107,10 @@ func TestNewContainer(t *testing.T) {
 	t.Parallel()
 
+	if runtime.GOOS == "darwin" {
+		t.Skip("test with /bin/sh isn't supported on darwin")
```
Why? Is this related to tty stuff?

thehajime replied:
Tty is also missing, in addition to the same reason to the above comment.

AkihiroSuda commented on container_test.go:
```diff
@@ -1174,6 +1234,10 @@ func TestContainerHostname(t *testing.T) {
 	t.Parallel()
 
+	if runtime.GOOS == "darwin" {
+		t.Skip("Test with true command doesn't run on darwin")
```
Why?

thehajime replied:
Ditto.

mxpv commented on container_test.go:
```diff
@@ -1783,6 +1864,10 @@ func TestContainerExecLargeOutputWithTTY(t *testing.T) {
+	if runtime.GOOS == "darwin" {
+		t.Skip("test with true command isn't supported on darwin")
+	}
+
```
Should we just exclude this file on darwin? Seems to be a lot cleaner solution.

thehajime replied:
thanks for the comment.

As I commented at https://github.com/containerd/containerd/pull/4526#discussion_r494017023, there are still valid tests in `container_test.go` even with darwin and it's worthwhile to have those tests I believe.

Excluding the tests is indeed much cleaner but I'd like to test if it's fine.

thehajime commented:
I have (force) pushed the branch to fix several issues.
There is still a blocker of go-runc upstream (https://github.com/containerd/go-runc/issues/65), which should be fixed asap (thus, `../project/script/validate/vendor` still reports an error).

thehajime commented:
Have pulled go-runc, which resolves the blocker (https://github.com/containerd/go-runc/issues/65), and rebased.

AkihiroSuda commented on platforms/defaults_darwin.go:
```diff
@@ -0,0 +1,31 @@
+func Default() MatchComparer {
+	return Ordered(DefaultSpec(), specs.Platform{
+		OS:           "linux",
+		Architecture: "amd64",
```
Should check GOARCH?

thehajime replied:
if you had apple silicon in your mind, then, yes.  it should be like this.

```suggestion
		Architecture: runtime.GOARCH,
```

(I updated the patch)

Is that what you're commenting about ?

AkihiroSuda replied:
yes

AkihiroSuda commented on platforms/defaults_darwin.go:
```diff
@@ -0,0 +1,31 @@
+// Default returns the default matcher for the platform.
+func Default() MatchComparer {
+	return Ordered(DefaultSpec(), specs.Platform{
+		OS:           "linux",
```
Please add a comment line to avoid confusion

thehajime replied:
added a comment

AkihiroSuda reviewed (approved):
Thanks

mxpv commented on helpers_darwin_test.go:
```diff
@@ -0,0 +1,61 @@
+}
+
+func withCat() oci.SpecOpts {
+	// TODO: use appropriate command instead of dummy (hello) command
```
nit: Remove TODO?

thehajime replied:
thanks, fixed in multiple places.

mxpv commented on helpers_darwin_test.go:
```diff
@@ -0,0 +1,61 @@
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+const newLine = "\n"
```
nit: Not used?

thehajime replied:
this variable is used at container_test.go (TestContainerOutput).

mxpv commented on .github/workflows/ci.yml:
```diff
@@ -455,3 +455,93 @@ jobs:
+      - name: Install Go
+        uses: actions/setup-go@v1
+        with:
+          go-version: '1.13.15'
```
Our CI is currently on `1.15.2`

thehajime replied:
thanks, fixed.

mxpv commented on .github/workflows/ci.yml:
```diff
@@ -455,3 +455,93 @@ jobs:
+          bindir="$(pwd)/bin/"
+          cd src/github.com/ukontainer/runu
+          git fetch --tags origin "${SHIM_COMMIT}"
+          git checkout "${SHIM_COMMIT}"
```
Shouldn't `go get github.com/ukontainer/runu@"${SHIM_COMMIT}"` already pull proper revision?

thehajime replied:
go get with `GO111MODULE=on` doesn't pull source, so we used previously checkouted tree to build the shim binary.
additionally, currently containerd-shim-darwin-v1 can't build with `GO111MODULE=on go get` so, I didn't touch this part.

thehajime replied:
okay; I will use go build to build runu binary.

mxpv commented on .github/workflows/ci.yml:
```diff
@@ -455,3 +455,93 @@ jobs:
+        shell: bash
+        run: |
+          set -o xtrace
+          cd src/github.com/containerd/containerd
```
nit: `working-directory: src/github.com/containerd/containerd`
(same below and above)

thehajime replied:
added multiple `working-directory`.
defaluts helps indeed, but some steps before checkouting containerd will fail so, I didn't.

thanks.

thehajime replied:
I was wrong; If the checkouting containerd is the earliest, I can use the `defaults` clause to simplify.  I've updated the ci.yml (and rebased).

mxpv commented on .github/workflows/ci.yml:
```diff
@@ -455,3 +455,93 @@ jobs:
+          # /run is read-only under macOS
+          export EXTRA_TESTFLAGS="-address=$WORKDIR/c1.sock"
+          export TESTFLAGS_PARALLEL=1
+          export PATH=$PATH:$(go env GOPATH)/bin
```
This can be simplified:
`export PATH` is not needed, since you already setup PATH from `Set env` above.
Same for `Integration 1`.

thehajime replied:
fixed, thanks.

mxpv reviewed (commented):
Thanks for the PR!
I have a few suggestions to simplify CI and cleanup related nits.

thehajime commented:
I rebased the branch to the current master, and added `args: --timeout=300s` to golangci-lint parameter because recent changes makes the check longer.

AkihiroSuda commented on defaults/defaults_darwin.go:
```diff
@@ -0,0 +1,35 @@
+	// to store FIFOs.
+	DefaultFIFODir = "/var/run/containerd/fifo"
+	// DefaultRuntime is the default darwin runtime
+	DefaultRuntime = "io.containerd.darwin.v1"
```
There can be multiple runtime implementations for Darwin, so maybe we should keep `DefautlRuntime` empty and call it a day to get this PR merged early

thehajime replied:
fixed.  thanks for the comment.

AkihiroSuda replied:
The value is still non-empty

thehajime replied:
Oops, my fault..  Updated again.

thehajime commented:
rebased the latest master, applying new vendor way, updates ci.yml, resolves conflicts on shim, etc.
met a CI failure while re-rebasing to the master; I will investigate what's going on..

hazcod commented:
@thehajime thank you for this PR!
FYI there is one conflict remaining.

thehajime commented:
@hazcod thanks for the head up 👍 
I've fixed the conflict: the new CI error seems to not be related to this PR.

kjenney commented:
@thehajime What's the status on this? Do you need any help with something specific?

thehajime commented:
> @thehajime What's the status on this? Do you need any help with something specific?

I'm always wishing to upstream this patchset, waiting for responses from maintainers.

I will work on to resolve the conflicts asap.

kzys commented:
While runu is really, really cool, I'm hesitant to say runu is *the* macOS support, but okay to have that as *a* macOS support. We have made containerd somewhat working on FreeBSD (#4822 and #4858) while [runj](https://github.com/samuelkarp/runj) is still experimental. runu and macOS support could be the same. We have also made a change to run all unit tests on macOS (see #5264).

@crosbymichael @mxpv What do you folks think? You may know macOS better than the rest of us ;)

mxpv commented:
That make sense to me, I definitely want compatibility patches in for darwin. As for the runtime/runu part - I'd suggest to start with a non core repo to give things a try first.

thehajime commented:
@kzys @mxpv 

> While runu is really, really cool, I'm hesitant to say runu is _the_ macOS support, but okay to have that as _a_ macOS support.

I'm also with both of you; runu is an example of OCI runtimes which run on macOS.  There are many places in this PR which assume to use runu (especially in test cases), but the runtime and its correspondent shim code are not included in this PR.

Thus if there are new runtimes running on macOS, this patch is useful even if the runtime is not runu.

thehajime commented:
rebased, no essential changes are introduced.  
And it also adapted to new macOS unit tests of the CI file (ci.yml).

Any suggestions/comments are appreciated !

kzys commented on .github/workflows/ci.yml:
```diff
@@ -475,7 +479,63 @@ jobs:
+
+      - name: Integration 1
+        env:
+          TEST_RUNTIME: "io.containerd.darwin.v1"
```
How about using ukcontainer or runu's namespace instead of `io.containerd`? For example firecracker-containerd is under `aws.firecracker`.

thehajime replied:
https://github.com/containerd/containerd/pull/4526#discussion_r482694602

I received a similar comment before but didn't fix it; I'll change this to 'io.containerd.runu.v1' based on the previous comment.

kzys commented on Makefile.darwin:
```diff
@@ -16,7 +16,11 @@
 
-# amd64 supports go test -race
+# amd64 supports go test -race,
+# but causes race-limit of simultaneously goroutines, thus disabled
```
I'd like to keep `-race` here for catching concurrency issues. What is the issue you'd like to workaround?

thehajime replied:
Agree.

I'm sure that this caused the issue of something related to goroutine rate-limit, but now I cannot reproduce this anymore in several cases (use old my PR patchset, downgrade go version, etc).

I'll revert this changes to the original one.

Thanks for the head up.

kzys commented on archive/tar_test.go:
```diff
@@ -747,6 +748,13 @@ func TestBreakouts(t *testing.T) {
 	}
 
 	for _, bo := range breakouts {
```
I'm sorry if you have answered about that already, but why? Hardlinks themselves are supported in macOS.

thehajime replied:
right, other Hardlink-related tests are fine actually.

I found that now this test won't run with the PR below, so I'll revert this diff for the next.

https://github.com/containerd/containerd/pull/5264#discussion_r601628758

kzys commented on integration/client/helpers_darwin_test.go:
```diff
@@ -0,0 +1,58 @@
+
+const newLine = "\n"
+
+func withExitStatus(es int) oci.SpecOpts {
+	return func(_ context.Context, _ oci.Client, _ *containers.Container, s *specs.Spec) error {
+		// TODO: use appropriate command instead of dummy (hello) command
+		s.Process.Args = []string{"hello"}
+		return nil
+	}
+}
```
Why do we have this "hello" while some others tests are skipped?

thehajime replied:
because currently available runtime (i.e., runu) only supports a single process, there are several limitations on various tests.

If a test only involves a single process (doesn't use fork syscall), then this hello program is used.

kzys commented on platforms/defaults_darwin.go:
```diff
@@ -0,0 +1,34 @@
+)
+
+// Default returns the default matcher for the platform.
+func Default() MatchComparer {
+	return Ordered(DefaultSpec(), specs.Platform{
+		// darwin runtime also supports Linux binary via runu/LKL
```
Isn't the runtime *only* supporting Linux binary?

thehajime replied:
not only.

currently the runtime support Linux/ELF and macOS/Mach-O (GOOS=darwin) binaries.

cpuguy83 replied:
We should be careful here. `runu` is not native.

thehajime commented:
rebased again with fixes proposed by @kzys (thanks!).

kzys reviewed (approved):
LGTM!

@mxpv Can you take a look?

AkihiroSuda commented:
@samuelkarp PTAL?

AkihiroSuda commented on integration/client/client_darwin_test.go:
```diff
@@ -0,0 +1,43 @@
+)
+
+func init() {
+	testImage = "docker.io/ukontainer/runu-base:0.3-osx-extra"
```
Can we pass the image string via GitHub Actions yaml?

I don't think we have reached consensus on having runU dependency on Go codes. (Neither on yamls, but yamls might be easier than Go codes to reach consensus)

AkihiroSuda replied:
Also,  we want to avoid using Docker Hub, unless they allow limitless pulling for GHA macOS instances.

thehajime replied:
> Also, we want to avoid using Docker Hub, unless they allow limitless pulling for GHA macOS instances.

i'm not familiar with alternatives.  do you have any recommendation for an alternate repository ?

AkihiroSuda replied:
ghcr.io seems good

thehajime replied:
thanks, fixed as you suggested, using ghcr instead of docker hub.

estesp reviewed (approved):
LGTM

mikebrow commented on sys/reaper/reaper_unix.go:
```diff
@@ -58,9 +58,10 @@ func (s *subscriber) do(fn func()) {
+func Reap(bundlePath string) error {
 	now := time.Now()
-	exits, err := reap(false)
+	exits, _ := reap(false)
```
for non darwin.. it looks like this just forces Reap to return nil no matter if an error occurred in reap()

thehajime replied:
Thanks, you're right.  It should try to preserve previous behavior.
I'll fix this in next patch.

mikebrow commented on .github/workflows/ci.yml:
```diff
@@ -479,14 +479,18 @@ jobs:
+    name: MacOS integration
     runs-on: macos-10.15
-    timeout-minutes: 10
+    timeout-minutes: 30
```
sounds high?

mikebrow commented on Makefile.darwin:
```diff
@@ -20,3 +20,6 @@ COMMANDS += containerd-shim
 endif
+
+# cri is not implemented yet on darwin
+SKIPTESTS := github.com/containerd/containerd/pkg/cri/server
```
newline..

thehajime replied:
I will fix it too.  Thanks.

thehajime commented:
rebased and addressed comments by @mikebrow.

samuelkarp commented on sys/reaper/reaper_darwin.go:
```diff
@@ -0,0 +1,60 @@
+// reapMore polls the status of registered process and terminate it
+// if it's already exited.
+
+func reapMore(bundlePath string, exits []exit) ([]exit, error) {
+	pid, err := runc.ReadPidFile(filepath.Join(bundlePath, process.InitPidFile))
```
`bundlePath` can be `""`

```suggestion
func reapMore(bundlePath string, exits []exit) ([]exit, error) {
	if bundlePath == "" {
		return nil, nil
	}
	pid, err := runc.ReadPidFile(filepath.Join(bundlePath, process.InitPidFile))
```

thehajime replied:
I investigated this again and found that the parameter `bundlePath` is not needed to pass (it can be estimated from the current directory) so, reverted the related changed to the original one.

samuelkarp commented on runtime/v2/runc/container.go:
```diff
@@ -139,24 +137,7 @@ func NewContainer(ctx context.Context, platform stdio.Platform, r *task.CreateTa
-			}
-		}
-		container.cgroup = cg
+		newContainerCgroup(pid, container)
```
Are these changes necessary since runu has an out-of-tree shim?  The runj shim doesn't import this package, for example.

thehajime replied:
I'm going to investigate this too if we can avoid importing (not done yet), but I guess with this diff the out-of-tree shim can reduce the duplicated portion of code.

thehajime replied:
Now I spend some time to make this hunk of patch eliminated and successfully removed from this PR by moving portion of code to (out-of-tree) shim implementation.
(I referred the runj shim code).

So this diff is no longer exist in this patchset.  Thanks for the suggestion.

samuelkarp commented on runtime/v2/shim/shim_darwin.go:
```diff
@@ -18,7 +18,9 @@
 package shim
 
-import "github.com/containerd/ttrpc"
+import (
+	"github.com/containerd/ttrpc"
+)
```
Can you revert this?

thehajime replied:
great catch! will fix this.

Thanks.

kzys commented on defaults/defaults_darwin.go:
```diff
@@ -0,0 +1,37 @@
+	// DefaultRootDir is the default location used by containerd to store
+	// persistent data
+	DefaultRootDir = "/var/lib/containerd"
+	// DefaultStateDir is the default location used by containerd to store
+	// transient data
+	DefaultStateDir = "/var/run/containerd"
+	// DefaultAddress is the default unix socket address
+	DefaultAddress = "/var/run/containerd/containerd.sock"
+	// DefaultDebugAddress is the default unix socket address for pprof data
+	DefaultDebugAddress = "/var/run/containerd/debug.sock"
+	// DefaultFIFODir is the default location used by client-side cio library
+	// to store FIFOs.
+	DefaultFIFODir = "/var/run/containerd/fifo"
```
Yes. It doesn't. I think `/run` is Linux thing.

https://lwn.net/Articles/436012/

https://www.freebsd.org/cgi/man.cgi?hier(7)

thehajime replied:
Also more seriously, `/run` is not writable on darwin without SIP (System Integrity Protection) disabled even if having root privilege (AFAIK).

thehajime commented on sys/reaper/reaper_darwin.go:
```diff
@@ -0,0 +1,60 @@
+			Status: 0, // XXX
+		})
+
+		logrus.Warnf("reapMore: detect exited, pid=%d", pid)
```
Thanks, will fix this.

thehajime commented on sys/reaper/reaper_darwin.go:
```diff
@@ -0,0 +1,60 @@
+	if err != nil && strings.Contains(err.Error(), "os: process already finished") {
+		exits = append(exits, exit{
+			Pid:    pid,
+			Status: 0, // XXX
```
> Is this still TODO?

Yes.  Because macOS doesn't have subreaper implementation in kernel (PR_SET_CHILD_SUBREAPER on Linux, PROC_REAP_ACQUIRE on FreeBSD, etc), this is way downgraded procedure to reap exited process by polling the status.

This is not an ideal portion of darwin port, but cannot find a nicer way to work around.

thehajime commented on sys/reaper/reaper_darwin.go:
```diff
@@ -0,0 +1,60 @@
+// Since macOS doesn't raise SIGCHLD on orphaned children's exit,
+// reapMore polls the status of registered process and terminate it
+// if it's already exited.
+
```
Thx, will fix this too.

thehajime commented on sys/reaper/reaper_darwin.go:
```diff
@@ -0,0 +1,60 @@
+	process, err := os.FindProcess(pid)
+	// ensure the process is running
+	if process != nil {
+		err = process.Signal(syscall.Signal(0))
```
agree. will add a comment.

thehajime commented on sys/reaper/reaper_unix.go:
```diff
@@ -58,9 +58,12 @@ func (s *subscriber) do(fn func()) {
 	now := time.Now()
 	exits, err := reap(false)
+	if err != nil {
+		exits, err = reapMore(bundlePath, exits)
```
agree. will fix this.

thehajime commented:
force-pushed with fixed several @samuelkarp 's suggestions.

gwuah commented:
Hi, what's the update on this pull request? 🙏

cndoit18 commented on runtime/v2/runc/container_darwin.go:
```diff
@@ -0,0 +1,27 @@
+	"github.com/containerd/containerd/pkg/process"
+)
+
+func newContainerCgroup(pid int, container *Container) {
```
Is there any solution for this piece now?

thehajime replied:
I worked and this code is no longer in this patchset.

cndoit18 commented:
A great feature, I'm eager to run it on my mac

thehajime commented:
I have rebased to _main_ branch and addressed some cleanup for shim implementation, reducing the diffs for this PR.  I'm wondering what is the right step to move forward this pull requests.

slonopotamus commented:
Possibly a dumb question: what kind of containers this PR will add support for?

I read that [runu](https://github.com/ukontainer/runu) says it is "OCI runtime for frankenlibc unikernel".

In turn, [frankenlibc](https://github.com/justincormack/frankenlibc) is "Frankenlibc is a build tool to build applications with userspace rump kernels based unikernels, as well as for some bare metal platforms. Currently there are three userspace implementations included, NetBSD, Linux and FreeBSD".

Do I understand correctly that this PR will allow containerd to run FreeBSD (or NetBSD???) as an "unikernel" on top of darwin in a kind of container (that is somehow different from a virtual machine, but I'm not sure how exactly)?

samuelkarp commented:
Hey @thehajime, sorry for the slow response.

> I'm wondering what is the right step to move forward this pull requests.

I want to be clear that I'm only speaking for myself here, not for the other maintainers.  However, I have a few concerns with the PR as-is and I don't really know what the right answer is yet.

1. I'm not entirely sure that it makes sense to call this "darwin runtime support".  runu is pretty neat, but as far as I can tell frankenlibc programs are not "native darwin" programs.
2. There appear to be some fundamental gaps in darwin's ability to provide some of facilities containerd needs.  For example, the symlink-based "bind-mount" approach is an interesting way to get around the lack of bind-mounts or nullfs mounts, but doesn't provide the same kind of filesystem isolation.  Subreaping is a similar gap; the approach used can either provide no data (`Status: 0` for an already-exited process) or get incorrect data if the PID gets recycled.
3. If we're going to add CI based on runu we'd like to either make sure it's really solid or make sure we can skip considering it if flakes occur.  Can you commit to helping us out if runu causes CI failures?
4. Lots of tests are currently skipped for darwin and runu.  Are those tests you'd plan to enable in the future by enhancing runu to support those use-cases?
5. There are lots of TODOs all over the codebase.  I don't think everything needs to be done at the same time, but I'd prefer to see functionality omitted entirely rather than stubbed out with TODO comments.

It might be worth trying to get this done as a series of smaller, more focused PRs.  This is the route I was going down with FreeBSD support and runj.  For example:

* A PR just focused on fixing any build errors (not sure if this applies, since I think we already build for darwin).  I did this for FreeBSD in https://github.com/containerd/containerd/pull/4822.
* A PR just focused on the `mount` changes and getting a snapshotter to work.  https://github.com/containerd/containerd/pull/4858 was my approach leveraging nullfs for FreeBSD.  Given darwin's lack of bind-mount/nullfs support I don't know what approach should be taken, but at least moving those changes into their own PR decouples those concerns from anything else.
* A PR just focused on enabling runtimes on darwin.  https://github.com/containerd/containerd/pull/5375 was what I did for FreeBSD.  Like the concern about mount support, the lack of native subreaping might be an issue.  However, the subreaping _is_ just a function of the shim, so the existing approach you have could be moved into the runu shim rather than being pushed into containerd itself.  I've done something similar in the [runj shim](https://github.com/samuelkarp/runj/blob/main/containerd/reaper.go).  I have my own code for setting up `_PROC_REAP_ACQUIRE` and then invoke `reaper.Reap()` myself; you could have your own reaper implementation specific to runu instead.
* A PR just focused on tests.  I haven't done this for FreeBSD yet, but it's on the list of things I'd like to do.
* A PR just focused on enabling CI, once everything else is done.

I hope this is helpful even though it's not a simple answer.  I want to reiterate that this is just my opinion rather than a consensus of the other maintainers too.

And finally: I'm really excited to see containerd support come to other platforms, so I'm rooting for this to be a success.  Thanks for putting everything together and my apologies for how slow and arduous the process has been so far.

thehajime commented:
> I want to be clear that I'm only speaking for myself here, not for the other maintainers. However, I have a few concerns with the PR as-is and I don't really know what the right answer is yet.

Thanks for your valuable comments.  It is always helpful to me.

> 1. I'm not entirely sure that it makes sense to call this "darwin runtime support".  runu is pretty neat, but as far as I can tell frankenlibc programs are not "native darwin" programs.

frankenlibc just offers a wrapper for cross compilation.  it also can generate mach-o binaries.
Plus, this containerd patches can play with other runtimes running on macOS/darwin so, I put the name of this PR as 'darwin runtime support'.

> 2. There appear to be some fundamental gaps in darwin's ability to provide some of facilities containerd needs.  For example, the symlink-based "bind-mount" approach is an interesting way to get around the lack of bind-mounts or nullfs mounts, but doesn't provide the same kind of filesystem isolation.  Subreaping is a similar gap; the approach used can either provide no data (`Status: 0` for an already-exited process) or get incorrect data if the PID gets recycled.

I agree.
Outsourcing reaper implementation to shim should be appropriate.  I will also work on this.

> 3. If we're going to add CI based on runu we'd like to either make sure it's really solid or make sure we can skip considering it if flakes occur.  Can you commit to helping us out if runu causes CI failures?

I thought I'll be responsible at all the related containerd code which my commits involve so, yes, I'll do that.

> 4. Lots of tests are currently skipped for darwin and runu.  Are those tests you'd plan to enable in the future by enhancing runu to support those use-cases?

Multiple processes support is still under the way and not likely to come very soon.
But if it is coming, I can remove those skips in the future (but cannot tell when it is).

> 5. There are lots of TODOs all over the codebase.  I don't think everything needs to be done at the same time, but I'd prefer to see functionality omitted entirely rather than stubbed out with TODO comments.
> 
> It might be worth trying to get this done as a series of smaller, more focused PRs. This is the route I was going down with FreeBSD support and runj. For example:
> 
> * A PR just focused on fixing any build errors (not sure if this applies, since I think we already build for darwin).  I did this for FreeBSD in [Build on FreeBSD #4822](https://github.com/containerd/containerd/pull/4822).
> * A PR just focused on the `mount` changes and getting a snapshotter to work.  [Support the native snapshotter on FreeBSD #4858](https://github.com/containerd/containerd/pull/4858) was my approach leveraging nullfs for FreeBSD.  Given darwin's lack of bind-mount/nullfs support I don't know what approach should be taken, but at least moving those changes into their own PR decouples those concerns from anything else.
> * A PR just focused on enabling runtimes on darwin.  [freebsd: runtime support #5375](https://github.com/containerd/containerd/pull/5375) was what I did for FreeBSD.  Like the concern about mount support, the lack of native subreaping might be an issue.  However, the subreaping _is_ just a function of the shim, so the existing approach you have could be moved into the runu shim rather than being pushed into containerd itself.  I've done something similar in the [runj shim](https://github.com/samuelkarp/runj/blob/main/containerd/reaper.go).  I have my own code for setting up `_PROC_REAP_ACQUIRE` and then invoke `reaper.Reap()` myself; you could have your own reaper implementation specific to runu instead.
> * A PR just focused on tests.  I haven't done this for FreeBSD yet, but it's on the list of things I'd like to do.
> * A PR just focused on enabling CI, once everything else is done.

Thanks for the great instruction.  I think splitting this PR into severals is a good idea and is the way to go.
I will try to reorganize things.

So far, I was following the way how Windows containerd support did; skipping tests to pass all, how CI should be looking like, etc.  But I should also consider FreeBSD (as it is similar).

> I hope this is helpful even though it's not a simple answer. I want to reiterate that this is just my opinion rather than a consensus of the other maintainers too.
> 
> And finally: I'm really excited to see containerd support come to other platforms, so I'm rooting for this to be a success. Thanks for putting everything together and my apologies for how slow and arduous the process has been so far.

Thanks, will do my best (tm).

thehajime commented:
Hi @slonopotamus,

> Possibly a dumb question: what kind of containers this PR will add support for?

this PR tries to run containerd process on darwin; it conceptually does not restrict the kind of containers running.  but it is now only tested with runu runtime.

> Do I understand correctly that this PR will allow containerd to run FreeBSD (or NetBSD???) as an "unikernel" on top of darwin in a kind of container (that is somehow different from a virtual machine, but I'm not sure how exactly)?

runu currently can run on Linux (x86/arm/aarch64) and macOS (x86), with the custom image generated with frankenlibc, and the official alpine linux image on Linux and macOS (with limited functionality support).

beveradb commented:
> I will try to reorganize things.

Hey @thehajime, do you need any help with this? Or, is there anything external third parties can do to support?

I'd love to see this PR move forward, especially since Docker Desktop (with it's convenient hyperkit VM which has provided the default local container runtime for so many devs these days) [will now charge a fee](https://www.google.com/search?q=docker+desktop+news&tbm=nws) for professional use from January.

We're evaluating whether the most pragmatic solution is to invest efforts into OSS alternatives which still rely on a VM (e.g. [podman](https://github.com/containers/podman) ), or whether it's realistic to get this approach working this year 😅

thehajime commented:
> Hey @thehajime, thank you so much for your efforts to try and get this working!
> Do you need any help with the work involved in addressing @samuelkarp's comments? Or, is there anything external third parties can do to support?

Thanks for reminding me ! The delay is just a matter of my laziness to allocate my time.

I'm starting to work on the reorganization, and will publish soon.  Hopefully it will help a bit.



> I'd love to see this PR move forward, especially since Docker Desktop (with it's convenient hyperkit VM which has provided the default local container runtime for so many devs these days) [will now charge a fee](https://www.google.com/search?q=docker+desktop+news&tbm=nws) for professional use from January.

I hope this bit is a step toward an alternate docker desktop (on mac), though I see this would require more effort not the only container runtime side, but the kernel part (i.e., linux kernel library/LKL) which I'm also working on.

thehajime commented:
I made a chunk of PRs from this patchset:

1. snapshotter support (#5935)
1. runtime support (#5936)
1. CI tests (not yet)

Once first 2 PRs are done, I will follow up with the last PR.

n0rig commented:
@thehajime this is really great! I am curious if you have an eta for merging the snapshotter pr

cosmojg commented:
Is this pull request going to be merged? Or is it obsolete?

jadnhm commented:
> Is this pull request going to be merged? Or is it obsolete?

I think what's happened is that the main implementation of  this has been put into 2 different PRs, 1 of which has been merged but the other "snapshotter" hasn't gained approval and hasn't been merged
https://github.com/containerd/containerd/pull/4526#issuecomment-912185193

@thehajime put a lot of time into this but I feel like he didn't receive the necessary support/attention to get this through in a reasonable amount of time and he's likely 'moved on' with his life a bit haha 😝 

I think he's still active in LKL-land so maybe this hasn't fallen off his radar completely. 

It seems to me like macOS/darwin does have lots of tech to lean on to get something like this working and there are lots of supporting OSS projects that could be brought to bear also (gvisor, lkl, etc) also.  

It would be truly amazing to have a library-based (ie: no VMs) container runtime for the platform.

thehajime commented:
> Is this pull request going to be merged? Or is it obsolete?

as @jadnhm said all.

the only remained item is the snapshotter PR, which hasn't moved forward for a while.  While my life still has a room to work on this, I'm waiting something happens for this.

thanks for the message.

Privatecoder commented:
lets please not let this die..

jadnhm commented:
@thehajime 

Can we infer from this comment that the snapshotter part of this might not even be necessary anymore?
https://github.com/containerd/containerd/discussions/5525#discussioncomment-6298614

slonopotamus commented:
#8789 was reverted, and containerd is still missing ability to mount/unmount things on macOS. That's why I have to apply patches in https://github.com/macOScontainers/containerd. Sparsebundle-based suggested by @thehajime in #5935 might be superior to native snapshotter in terms of performance. But both things got stuch because containerd devs want a pluggable solution that doesn't exist currently. Though for some reason that doesn't stop #8807 from being merged, what I treat as double standards.

AkihiroSuda commented:
> Though for some reason that doesn't stop https://github.com/containerd/containerd/pull/8807 from being merged, what I treat as double standards.

The story is quite different for cimfs (#8807), as it is part of the operating system.

We could probably reach the consensus on merging https://github.com/containerd/containerd/pull/8789 if bindfs were part of macOS.

dmcgowan commented:
> Sparsebundle-based suggested by @thehajime in https://github.com/containerd/containerd/pull/5935 might be superior to native snapshotter in terms of performance. But both things got stuck because containerd devs want a pluggable solution that doesn't exist currently.

I don't think the helper binaries with names resolved by the mount type is the right plugin interface to support custom mount types. The goal of a plugin interface is that they would be easy to register and work with other plugins, such as a snapshotter that uses a custom mount type and able to register the custom mount type. Also such a plugin interface needs to worry about lifecycle and device activation which is not relevant for system mount types. 

I do think we could have a solution which is less complicated though, we should have a support for mounting block devices with natively supported filesystems using hdiutil and bind mounts using bindfs (possibly able to be overwritten via environment but using bindfs cli interface). Currently the helpers just wrap these utilities but I don't see a reason why we can't use them directly in our mount package. Then we can defer the debate/development over custom mount types supported across all OSes.

kjenney commented:
Up to avoid stale.

9numbernine9 commented:
Keepalive bump!

xyb commented:
ensure this survives the bot

slonopotamus commented:
Almost everything from this PR was already merged in https://github.com/containerd/containerd/pull/5936

slonopotamus commented:
So, I think this PR can be closed. The only useful changes left here are disabling of tests. But until we have bind-mounts implementation, there's no way to actually run them. And all core changes were already merged.

slonopotamus commented:
Since Xcode 15, Apple adds FSKit in betas, and then removes in RC/release. Then adds back in next beta.

lelvisl commented:
mac os 15.4 released, is FSKit available? and what else blocking this PR?
